visualization/app/codeCharta/state/store/metricData/edgeMetricData/edgeMetricData.reducer.ts

```diff
@@ -0,0 +1,118 @@
+import { EdgeMetricDataAction, EdgeMetricDataActions, setEdgeMetricData } from "./edgeMetricData.actions"
+import { BlacklistItem, BlacklistType, Edge, EdgeMetricCount, EdgeMetricData } from "../../../../codeCharta.model"
+import { getVisibleFileStates } from "../../../../model/files/files.helper"
+import { CodeMapHelper } from "../../../../util/codeMapHelper"
+import { FileState } from "../../../../model/files/files"
+import { EdgeMetricDataService } from "./edgeMetricData.service"
+
+const clone = require("rfdc")()
+
+// Required for performance improvements
+export type NodeEdgeMetricsMap = Map<string, Map<string, EdgeMetricCount>>
+export let nodeEdgeMetricsMap: NodeEdgeMetricsMap
+
+export function edgeMetricData(state: EdgeMetricData[] = setEdgeMetricData().payload, action: EdgeMetricDataAction): EdgeMetricData[] {
+	switch (action.type) {
+		case EdgeMetricDataActions.SET_EDGE_METRIC_DATA:
+			return clone(action.payload)
+		case EdgeMetricDataActions.CALCULATE_NEW_EDGE_METRIC_DATA:
+			return clone(calculateMetrics(action.payload.fileStates, action.payload.blacklist))
```

Comment:
I tried to figure out what `calculateMetrics` does but it was not trivial enough to see if the data is already copied or not. I expect the clone not to be necessary here, but this should be looked at again.